### ISSUE TITLE
Point cloud embeddings overwrite text embeddings?

### ISSUE THREAD

**Issue by jereminuer** [col14m/cadrille#4]
Hey - not sure if I'm misinterpreting this, if so, would love an explanation and feel free to close the issue. But looks like the point cloud embeddings overwrite your other tokens.

In the collate function of `cadrille.py`:
```
    points_inputs = ''.join(n_points * [processor.tokenizer.pad_token])
    
    for i in range(len(texts)):
        if is_pc[i]:
            texts[i] = points_inputs + texts[i]

    image_inputs, video_inputs = process_vision_info(messages)
    inputs = processor(
        text=texts,
        images=image_inputs,
        videos=video_inputs,
        padding=True,
        return_tensors='pt')
```
This creates a list of repeated pad tokens, for the number of point_inputs, and prepends them to the actual message that is built. Then passes it through the processor to tokenize. This makes sense - create placeholders where the pointclouds can be inserted.

However, in the re-defined forward pass of `Qwen2VLForConditionalGeneration` of `cadrille.py`:
```
# add point cloud embeddings, we add only next 6 lines
if is_pc.sum() > 0 and (past_key_values is None or past_key_values.get_seq_length() == 0):
    point_embeds = self.point_encoder(point_clouds.float()).bfloat16()
    start_idxs = attention_mask.shape[1] - attention_mask.sum(axis=1) # (batch_size, )
    for i, start_idx in enumerate(start_idxs):
        if is_pc[i]:
            inputs_embeds[i, start_idx:start_idx + point_embeds.shape[1], :] = point_embeds[i]

if attention_mask is not None:
    attention_mask = attention_mask.to(inputs_embeds.device)
```

In the line:
```
start_idxs = attention_mask.shape[1] - attention_mask.sum(axis=1)
```
you compute indexes by taking the overall length of sequences (after we've padded to the maximum sequence length), and subtracting it by the number of **non**-padded tokens. In other words, this will count the number of padded tokens.

However, there are padded tokens that are prepended (from inserting pointclouds) and appended (via the normal padding operation that happens when you set `padding=True` upon passing the texts to the processor.

So this start index will be some amount of tokens after the prepended pad tokens. Then, the following operation happens:
```
inputs_embeds[i, start_idx:start_idx + point_embeds.shape[1], :] = point_embeds[i]
```
For each batch, you write from start_idx to however many point_embeddings you have.
Instead of writing to the prepended pad tokens, you write to text tokens, overwriting commands, and leaving blank empty tokens at the front.

Please let me know if I am misunderstanding. It's possible this wouldn't have an outsized impact if the only tokens it overwrote were text tokens that didn't have a huge impact on the task. But worth noting.
